### PR TITLE
lda_core.cc SIMD refactoring

### PR DESCRIPTION
**Pull request by bscottm**
[Sorry if the change set looks huge, caused in large part by running the code through clang-format to make it more readable.]

Refactor the LDA inline SIMD code into two blocks: one block for fast approximations vs. accuracy, the second block for SIMD code vs. serial code.

Update x86 SIMD code and use SSE3, SSE4.2 intrinsics, if the compiler has them enabled. Vector element extraction is template code with a specialization for the 0'th element. All of this results in better instruction selection in clang, gcc. [GCC 4.6.2 supports SSE4.1 with the "-msse4.1" flag. Not sure if this is desired in the Travis builds.]

Also prefer STL algorithms in a few places where they make sense, show examples of how C++ lambdas work (it's a C++0x feature.)

### CODE REVIEW

**Comment by bscottm** [comment]
If Boost.SIMD ever becomes a real thing (or the Github repo actually has working code), that will be a far better option than the hand-rolled intrinsics. Until then... this refactoring does allow me to add ARM NEON intrinsics for that port.

**Comment by arielf** [comment]
Scott, this looks nice.  Just to make sure: have you verified we don't have a runtime regression from this change.  I recall in the past Paul Mineiro switched to using some Intel SIMD instructions and we got a pretty significant degradation in speed.  Granted, it may have been a different change. Thanks!

**Comment by bscottm** [comment]
I haven't been particularly rigorous about collecting LDA SIMD vs. serial timing, although I did notice that I went from ~11 seconds for the entire test suite to ~9 seconds.

Is there a reasonable "soak test" for LDA?

Regression: There's little or no checking to ensure that the underlying arrays are vector aligned, so the compilers have to be pessimistic about loads and stores. That's were I'd put my money if I were to guess the origins of any SIMD-related regressions.

My patch shouldn't be quite as pessimistic if "-msse4.2" is passed to gcc. There's more XMMS register pressure, but that's not as awful as memory pressure from unnecessary spills to convert from __m128 to an array of floats.

**Comment by bscottm** [comment]
There are a few more tweaks I could look into that deal with alignment. And one more union to remove.

BTW: The SIMD code is enabled by default in both the pre- and post- patch code. If there are any regressions, they'll show up fairly quickly.

**Comment by bscottm** [comment]
Ariel: I've hacked the code so that all three (fast approx, accurate and SIMD) can be chosen at runtime, with a little help from templates. No more ifdef-ing. All code is inlined. Added a flag, "--math-mode <fast-approx|approx|accurate|accuracy|simd" to choose which math is used by the LDA core code (could probably be generalized across all of VW.)

Interesting regression, but I'm not sure how far back it goes: SIMD and Boost functions produce the same output, modulo 1e-4/1e-5 precision errors. The fast approximations are off by a factor of 2. Am investigating.

**Comment by arielf** [comment]
Two small, perhaps obvious comments:

Timing on the whole test-suite is misleading.  The more "precision differences ignored" lines you have, the slower the run-time because the ~10x slower fuzzy-FP match is skipped if the original diff succeeds.   The best test for speed of a certain test is to `time` an individual run.

I'm fine with converting `ifdef`'s to runtime switches where the benefit is greater than the increase in complexity and code size.  These routines are very small but (for any architechure) one of the options is faster so it should be the default without burdening the user with the additional complexity.

In any case, many thanks for doing all this research on what is really the fastest, and for all the other deep dives into the code trying to make it better.

**Comment by JohnLangford** [comment]
Paul, can you take a look at this in detail?

-John

**Comment by bscottm** [comment]
@John, @pmineiro, @arielf:

I pushed a newer version of the code that tidies the SIMD refactoring up. I've found that 'git diff -w' is the best way to compare the versions, since there are a lot of white space changes.

You can select between the different modes via the command line flag "--math-mode [fast-approx|approx|precise|simd]", with SIMD being the default. They produce similar results, although the test case output is likely to need regenerating if this is incorporated into the release.

At some point in the near future (next two weeks), I'll go and regenerate the test output for the three different modes and corresponding outputs.

**Comment by bscottm** [comment]
@JohnLangford @pmineiro @arielf:

Looks like I managed to get past the Boost.Alignment issue. Test is failing because the output needs regeneration. Please have a look at the build output and let me know how to proceed.

**Comment by arielf** [comment]
Hi Scott,

Thanks again for the work.

Looking at the travis failure:  https://travis-ci.org/JohnLangford/vowpal_wabbit/builds/55075029

Test 17 `diff -u --minimal train-sets/ref/wiki1K.stderr stderr.tmp`
is failing due to a significant difference in output.

This is not a floaing-point precision difference but something fundamentally different.

e.g: average loss after 256 examples is much lower in the new code.

This may actually be good (lower loss) but it needs to be verified that the current implementation is the correct one.  It looks to me that one of the implementations must be incorrect given the big difference in
output.  The incorrect one _may_ be the existing (old) one, meaning we had some accuracy bug all along, but this is less likely.

The question is which of the 3 implementations USE_SIMD, USE_PRECISE, USE_FAST_APPROX is causing such a big diff and is it actually correct.

**Comment by arielf** [comment]
Sorry, I realize I may not have been clear enough.  I understand this change switches from the present default (fast approx?) to using SIMD.  The main point is I don't expect such a change to have such a *large* effect on the output.  Ergo there has to be a bug somewhere.  The main purpose of `make test` is to catch these cases which is what happened here.  Changing the reference output is easy, but is it the correct fix?

**Comment by bscottm** [comment]
At least it's a consistent "feature":

```
$ ../vowpalwabbit/vw -k --lda 100 --lda_alpha 0.01 --lda_rho 0.01 --lda_D 1000 -l 1 -b 13 --minibatch 128 --math-mode simd /Users/michel/play/vowpal_wabbit/test/train-sets/wiki256.dat
Num weight bits = 13
learning rate = 1
initial_t = 0
power_t = 0.5
using no cache
Reading datafile = /Users/michel/play/vowpal_wabbit/test/train-sets/wiki256.dat
num sources = 1
average  since         example        example  current  current  current
loss     last          counter         weight    label  predict features
10.149698 10.149698            1            1.0  unknown   0.0000      732
10.369990 10.590282            2            2.0  unknown   0.0000       27
10.325998 10.282005            4            4.0  unknown   0.0000       53
10.401808 10.477618            8            8.0  unknown   0.0000       60
10.356298 10.310789           16           16.0  unknown   0.0000       26
10.473010 10.589722           32           32.0  unknown   0.0000      125
10.474932 10.476854           64           64.0  unknown   0.0000      313
10.425374 10.375817          128          128.0  unknown   0.0000       50
9.620762 8.816150          256          256.0  unknown   0.0000       33

finished run
number of examples = 256
weighted example sum = 256.000000
weighted label sum = 0.000000
average loss = 9.620762
total feature number = 22158
```

```
$ ../vowpalwabbit/vw -k --lda 100 --lda_alpha 0.01 --lda_rho 0.01 --lda_D 1000 -l 1 -b 13 --minibatch 128 --math-mode fast-approx /Users/michel/play/vowpal_wabbit/test/train-sets/wiki256.datNum weight bits = 13
learning rate = 1
initial_t = 0
power_t = 0.5
using no cache
Reading datafile = /Users/michel/play/vowpal_wabbit/test/train-sets/wiki256.dat
num sources = 1
average  since         example        example  current  current  current
loss     last          counter         weight    label  predict features
10.149614 10.149614            1            1.0  unknown   0.0000      732
10.369901 10.590187            2            2.0  unknown   0.0000       27
10.325905 10.281909            4            4.0  unknown   0.0000       53
10.401716 10.477528            8            8.0  unknown   0.0000       60
10.356205 10.310694           16           16.0  unknown   0.0000       26
10.472919 10.589634           32           32.0  unknown   0.0000      125
10.474844 10.476768           64           64.0  unknown   0.0000      313
10.425280 10.375717          128          128.0  unknown   0.0000       50
9.620675 8.816070          256          256.0  unknown   0.0000       33

finished run
number of examples = 256
weighted example sum = 256.000000
weighted label sum = 0.000000
average loss = 9.620675
total feature number = 22158
```

```
$ ../vowpalwabbit/vw -k --lda 100 --lda_alpha 0.01 --lda_rho 0.01 --lda_D 1000 -l 1 -b 13 --minibatch 128 --math-mode precise /Users/michel/play/vowpal_wabbit/test/train-sets/wiki256.datNum weight
Num weight bits = 13
learning rate = 1
initial_t = 0
power_t = 0.5
using no cache
Reading datafile = /Users/michel/play/vowpal_wabbit/test/train-sets/wiki256.dat
num sources = 1
average  since         example        example  current  current  current
loss     last          counter         weight    label  predict features
10.149390 10.149390            1            1.0  unknown   0.0000      732
10.369712 10.590034            2            2.0  unknown   0.0000       27
10.325692 10.281672            4            4.0  unknown   0.0000       53
10.401501 10.477310            8            8.0  unknown   0.0000       60
10.355986 10.310471           16           16.0  unknown   0.0000       26
10.472720 10.589454           32           32.0  unknown   0.0000      125
10.474631 10.476542           64           64.0  unknown   0.0000      313
10.425073 10.375516          128          128.0  unknown   0.0000       50
9.620526 8.815979          256          256.0  unknown   0.0000       33

finished run
number of examples = 256
weighted example sum = 256.000000
weighted label sum = 0.000000
average loss = 9.620526
total feature number = 22158
```

**Comment by arielf** [comment]
Thanks!

This consistency lends credence to the hypothesis that the new implementations are correct, and we had a (much worse) implementation inaccuracy before.  The small differences (< 0.001) in FP output between the 3 implementations are about what I expect from "FP implementation differences".

I'm leaning towards fixing the reference output now :)

@pmineiro if you get a chance to confirm and add your vote, it would be great.

Thanks.

**Comment by bscottm** [comment]
@arielf:

I was also more "stringent" about namespace qualification, especially when using functions in the std:: namespace. There are header files that have "using namespace std" but there is potential pollution in the source code that overrides and therefore hide functions that std:: provides. Don't ask me for a specific instance off the top of my head, though -- although some of the "mydigamma" and "mylgamma" preprocessor definitions come to mind.

I specifically qualifiy calls to std::exp, std::fmin and std::pow to ensure that the correct function is being called. Mostly out of paranoia -- I've been burned one too many times by namespace issues.

**Comment by bscottm** [comment]
@arielf:

Actually, this may be the reason: The threshold for expdigammify/vexpdigammify and expdgammify_2/vexdigammify_2 are consistently 1.0e-6. The previous code used 1.0e-6 for expdigammify and expdigammify_2 (non-simd) vs. 1.0e-10 for vexpdigammify/vexpdigammify_2 (the simd code).

Not hard to tweak if there's an urgent desire to go back to the previous threshold (although that'd have to wait until week after next.)

**Comment by pmineiro** [comment]
We should test on a large amount of input.  When I first implemented this way back when, I had underflow/overflow issues that only surfaced on larger data sets.  It's hard for me to tell from the diff :), but it'd be nice to get some confidence that we haven't reintroduced them somehow.

**Comment by bscottm** [comment]
@pmineiro: Was that the reason for the fmin(1.e-10, ...) in vexpdigammify?

**Comment by pmineiro** [comment]
@bscottm: details are hazy, but that sounds plausible.

**Comment by bscottm** [comment]
@pmineiro: That explains the inconsistency. The fmax value for expdigammify_2 is 1e-6 and 1e-10 for vexpdigammify_2 in the old code. They are now both set to 1e-6.

Also, the diff is easier to read if you look at it using the unified view, not the split view.

**Comment by bscottm** [comment]
Adjusted the underflow threshold to 1.0e-10. Test results align as expected on my simple local Mac. Let's see what Travis CI has to say… :-)

**Comment by bscottm** [comment]
And that's a good sign. :-)

**Comment by bscottm** [comment]
Just for completeness for @arielf:

```
$ ../vowpalwabbit/vw -k --lda 100 --lda_alpha 0.01 --lda_rho 0.01 --lda_D 1000 -l 1 -b 13 --minibatch 128 --math-mode simd ../test/train-sets/wiki256.dat
Num weight bits = 13
learning rate = 1
initial_t = 0
power_t = 0.5
using no cache
Reading datafile = ../test/train-sets/wiki256.dat
num sources = 1
average  since         example        example  current  current  current
loss     last          counter         weight    label  predict features
10.149310 10.149310            1            1.0  unknown   0.0000      732
10.369817 10.590323            2            2.0  unknown   0.0000       27
10.325925 10.282034            4            4.0  unknown   0.0000       53
10.401768 10.477611            8            8.0  unknown   0.0000       60
10.356288 10.310807           16           16.0  unknown   0.0000       26
10.472954 10.589620           32           32.0  unknown   0.0000      125
10.474859 10.476765           64           64.0  unknown   0.0000      313
10.425314 10.375769          128          128.0  unknown   0.0000       50
10.005566 9.585817          256          256.0  unknown   0.0000       33

finished run
number of examples = 256
weighted example sum = 256.000000
weighted label sum = 0.000000
average loss = 10.005566
total feature number = 22158
```

```
$ ../vowpalwabbit/vw -k --lda 100 --lda_alpha 0.01 --lda_rho 0.01 --lda_D 1000 -l 1 -b 13 --minibatch 128 --math-mode approx ../test/train-sets/wiki256.dat
Num weight bits = 13
learning rate = 1
initial_t = 0
power_t = 0.5
using no cache
Reading datafile = ../test/train-sets/wiki256.dat
num sources = 1
average  since         example        example  current  current  current
loss     last          counter         weight    label  predict features
10.149236 10.149236            1            1.0  unknown   0.0000      732
10.369734 10.590232            2            2.0  unknown   0.0000       27
10.325835 10.281937            4            4.0  unknown   0.0000       53
10.401677 10.477518            8            8.0  unknown   0.0000       60
10.356195 10.310714           16           16.0  unknown   0.0000       26
10.472864 10.589532           32           32.0  unknown   0.0000      125
10.474771 10.476678           64           64.0  unknown   0.0000      313
10.425222 10.375674          128          128.0  unknown   0.0000       50
10.005482 9.585742          256          256.0  unknown   0.0000       33

finished run
number of examples = 256
weighted example sum = 256.000000
weighted label sum = 0.000000
average loss = 10.005482
total feature number = 22158
```

```
$ ../vowpalwabbit/vw -k --lda 100 --lda_alpha 0.01 --lda_rho 0.01 --lda_D 1000 -l 1 -b 13 --minibatch 128 --math-mode precise ../test/train-sets/wiki256.dat
Num weight bits = 13
learning rate = 1
initial_t = 0
power_t = 0.5
using no cache
Reading datafile = ../test/train-sets/wiki256.dat
num sources = 1
average  since         example        example  current  current  current
loss     last          counter         weight    label  predict features
10.148996 10.148996            1            1.0  unknown   0.0000      732
10.369544 10.590092            2            2.0  unknown   0.0000       27
10.325634 10.281723            4            4.0  unknown   0.0000       53
10.401487 10.477340            8            8.0  unknown   0.0000       60
10.356004 10.310521           16           16.0  unknown   0.0000       26
10.472686 10.589368           32           32.0  unknown   0.0000      125
10.474590 10.476495           64           64.0  unknown   0.0000      313
10.425045 10.375500          128          128.0  unknown   0.0000       50
10.005359 9.585672          256          256.0  unknown   0.0000       33

finished run
number of examples = 256
weighted example sum = 256.000000
weighted label sum = 0.000000
average loss = 10.005359
total feature number = 22158
```

**Comment by JohnLangford** [comment]
I merged this in.  

The one thing that I find funny here is the use of memcpy instead of union.  Is that really required?

**Comment by bscottm** [comment]
@JohnLangford:

Just got back from an OCONUS trip and was out of touch for the duration. The memcpy is more intentional than the union and should tell future developers what's going on -- bit conversion from 32-bit floating point to 32-bit integer for the purposes of bit banging/munging. memcpy also avoids alignment issues: the union relies on integer and float alignment being the same, which can be a poor assumption.

gcc and clang's optimizers are pretty smart about reducing memcpy from a function call to an instruction or two. I'll post the assembly if that would help.

**Comment by bscottm** [comment]
@JohnLangford: Been doing software development for a while and have seen many strange things, so I tend to code defensively.

BTW: There are other places in the lda_core.cc code where loops could be vectorized. I don't want to recreate what the Boost.SIMD library (if it ever comes to fruition), but there are other vectorization opportunities.
